crates/polars-core/src/chunked_array/ops/gather.rs

```diff
@@ -0,0 +1,167 @@
+use arrow::array::Array;
+use polars_error::{polars_bail, polars_ensure, PolarsResult};
+
+use crate::chunked_array::ops::{ChunkTake, ChunkTakeUnchecked};
+use crate::chunked_array::ChunkedArray;
+use crate::datatypes::{IdxCa, PolarsDataType, StaticArray};
+use crate::prelude::*;
+use crate::utils::index_to_chunked_index;
+
+impl<T: PolarsDataType, I: AsRef<[IdxSize]> + ?Sized> ChunkTake<I> for ChunkedArray<T>
+where
+    ChunkedArray<T>: ChunkTakeUnchecked<I>,
+{
+    /// Gather values from ChunkedArray by index.
+    fn take(&self, indices: &I) -> PolarsResult<Self> {
+        let len = self.len();
+        let all_valid = indices.as_ref().iter().all(|i| (*i as usize) < len);
+        polars_ensure!(all_valid, ComputeError: "invalid index in gather");
+
+        // SAFETY: we just checked the indices are valid.
+        Ok(unsafe { self.take_unchecked(indices) })
+    }
+}
+
+unsafe fn target_value_unchecked<'a, A: StaticArray>(
+    targets: &[&'a A],
+    idx: IdxSize,
+) -> A::ValueT<'a> {
+    let (chunk_idx, arr_idx) =
+        index_to_chunked_index(targets.iter().map(|a| a.len()), idx as usize);
+    let arr = targets.get_unchecked(chunk_idx);
+    arr.value_unchecked(arr_idx)
+}
+
+unsafe fn target_get_unchecked<'a, A: StaticArray>(
+    targets: &[&'a A],
+    idx: IdxSize,
+) -> Option<A::ValueT<'a>> {
+    let (chunk_idx, arr_idx) =
+        index_to_chunked_index(targets.iter().map(|a| a.len()), idx as usize);
```

Comment:
Not for this PR, but the implementation of this function is so terrible. I wonder if you have a better idea for this.

Reply:
I have an idea which uses branchless binary searches.